### PR TITLE
DATAMONGO-719 - Add index creation for dynamically named collections.

### PR DESCRIPTION
**Pull request by ngbalk**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAMONGO).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

**Comment by ngbalk** [comment]
CC: @waillee @tsoberg

**Comment by ngbalk** [comment]
@Waillee

**Comment by nhomble** [comment]
This is awesome!

**Comment by ngbalk** [comment]
@christophstrobl @odrotbohm Curious to get your feedback on this PR! Thanks!

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/index/MongoPersistentEntityIndexCreator.java [hunk 1]
```diff
@@ -50,12 +51,13 @@
...
  * @author Christoph Strobl
  * @author Mark Paluch
+ * @author Nick Balkissoon
  */
 public class MongoPersistentEntityIndexCreator implements ApplicationListener<MappingContextEvent<?, ?>> {
 
 	private static final Logger LOGGER = LoggerFactory.getLogger(MongoPersistentEntityIndexCreator.class);
 
-	private final Map<Class<?>, Boolean> classesSeen = new ConcurrentHashMap<Class<?>, Boolean>();
+	private final Table<Class, String, Boolean> collectionsSeen = HashBasedTable.create();
```
Comment (erichhsun) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/index/MongoPersistentEntityIndexCreator.java]: `HashBasedTable` is not thread safe; you may want to consider `Tables.synchronizedTable()`: https://google.github.io/guava/releases/23.0/api/docs/com/google/common/collect/Tables.html#synchronizedTable-com.google.common.collect.Table-

Also, need to be careful about what you are synchronizing on when reading and writing.

Likewise, you may want to add an additional test to verify the behavior when multiple threads are using this class. I don't have enough context to tell if this is a valid use case.

Comment (ngbalk) [reply spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/index/MongoPersistentEntityIndexCreator.java]: Great feedback - looks like spring-data-mongodb depends on an older version of guava that doesn't support the synchronizedTable() , so I went back to using a ConcurrentHashMap... curious to hear what you think.

**Comment by ngbalk** [comment]
@christophstrobl  please review this PR :) 
Related ticket: https://jira.spring.io/browse/DATAMONGO-719

**Comment by mp911de** [comment]
Closing as per comment in [DATAMONGO-719](https://jira.spring.io/browse/DATAMONGO-719) without merging.

**Pull request closed without merging**